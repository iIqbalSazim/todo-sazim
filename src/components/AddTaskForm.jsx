import {
  Button,
  Center,
  Modal,
  Select,
  SimpleGrid,
  Textarea,
  Title,
} from "@mantine/core";
import { DateInput } from "@mantine/dates";
import { useState } from "react";
import { formatDueDate, generateCurrentTimeAndDate } from "../helper/helper";
import { PRIORITY } from "../constants/constant";
import { PRIORITY_OPTIONS } from "../constants/constant";
import { COLORS } from "../constants/constant";

const AddTaskForm = ({ createNewTask, isOpen, closeModal }) => {
  const [newTask, setNewTask] = useState({
    id: Math.round(Math.random() * 1000000),
    description: "Task details not provided",
    priority: PRIORITY.LOW,
    isCompleted: false,
    createdAt: "",
    dueDate: new Date().toDateString().slice(0, 15),
  });

  const handleSubmit = (e) => {
    e.preventDefault();
    createNewTask(newTask);
    setNewTask({
      description: "",
      priority: "",
      isCompleted: false,
      createdAt: "",
      updatedAt: "",
      dueDate: "",
    });

    closeModal();
  };

  const setCreatedAt = () => {
    let date = new Date();
    let createdAt = generateCurrentTimeAndDate(date);
    setNewTask({ ...newTask, createdAt: createdAt });
  };

  return (
    <Modal opened={isOpen} onClose={() => closeModal()} centered>
      <form onSubmit={handleSubmit}>
        <SimpleGrid cols={1} p={"lg"}>
          <Title ta={"center"}>Create Todo</Title>
          <Textarea
            placeholder="Write your task"
            onChange={(e) =>
              setNewTask({ ...newTask, description: e.target.value })
            }
          />
          <Select
            placeholder="Set priority"
            defaultValue={PRIORITY.LOW}
            data={PRIORITY_OPTIONS}
            onSelect={(e) =>
              setNewTask({
                ...newTask,
                priority: e.target.value.toLowerCase(),
              })
            }
          />
          <DateInput
            minDate={new Date()}
            defaultValue={new Date(newTask.dueDate)}
            onChange={(input) => {
              setNewTask({ ...newTask, dueDate: formatDueDate(input) });
            }}
            label="Date input"
          />
          <Center>
            <Button color={COLORS.BUTTON} type="submit" onClick={setCreatedAt}>
              Create
            </Button>
          </Center>
        </SimpleGrid>
      </form>
    </Modal>
  );
};

export default AddTaskForm;
