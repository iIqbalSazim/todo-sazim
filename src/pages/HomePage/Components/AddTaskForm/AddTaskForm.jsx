import { useState } from "react";

import {
  Button,
  Center,
  Modal,
  Select,
  SimpleGrid,
  Textarea,
  Title,
} from "@mantine/core";
import { DateInput } from "@mantine/dates";

import { PRIORITY, PRIORITY_OPTIONS, COLORS } from "../../HomePageConstants";
import { createTask } from "../../Api/Methods";

const AddTaskForm = ({ isOpen, closeModal }) => {
  const [newTask, setNewTask] = useState({
    description: "Task details not provided",
    priority: PRIORITY.LOW,
    is_completed: false,
    due_date: new Date().toDateString(),
  });

  const handleSubmit = async (e) => {
    e.preventDefault();

    await createTask({ task: { ...newTask } });

    closeModal();
    window.location.reload(false);
    alert("Task successfully added");
  };

  return (
    <Modal opened={isOpen} onClose={() => closeModal()} centered>
      <form onSubmit={handleSubmit}>
        <SimpleGrid cols={1} p={"lg"}>
          <Title ta={"center"}>Create Todo</Title>
          <Textarea
            placeholder="Write your task"
            onChange={(e) =>
              setNewTask({ ...newTask, description: e.target.value })
            }
          />
          <Select
            placeholder="Set priority"
            defaultValue={PRIORITY.LOW}
            data={PRIORITY_OPTIONS}
            onSelect={(e) =>
              setNewTask({
                ...newTask,
                priority: e.target.value.toLowerCase(),
              })
            }
          />
          <DateInput
            minDate={new Date()}
            defaultValue={new Date()}
            onChange={(input) => {
              setNewTask({ ...newTask, due_date: input });
            }}
            label="Due Date"
          />
          <Center>
            <Button color={COLORS.BUTTON} type="submit">
              Create
            </Button>
          </Center>
        </SimpleGrid>
      </form>
    </Modal>
  );
};

export default AddTaskForm;
