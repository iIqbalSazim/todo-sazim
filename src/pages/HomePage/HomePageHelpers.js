export function findIndexWithId(data, id) {
  return data.findIndex((element) => element.id === id);
}

export function formatCompletedAt(dateString) {
  const date = new Date(dateString);

  let currentTime =
    date.toLocaleTimeString().slice(0, 4) + date.toLocaleTimeString().slice(7);
  let currentDate = `${currentTime} | ${date.toDateString()}`;
  return currentDate;
}

export function formatDueDate(dateString) {
  const date = new Date(dateString);
  return date.toDateString().slice(0, 15);
}

export const replaceTabId = (htmlString) => {
  return htmlString.replace(
    /id="mantine-[a-z0-9]+-tab-(all|active|completed|target)"/g,
    (match, p1) => `id="mantine-mockedId-tab-${p1}"`
  );
};

export const replaceMenuTargetId = (htmlString) => {
  return htmlString.replace(
    /id="mantine-[a-z0-9]+-(target)"/g,
    (match, p1) => `id="mantine-mockedId-${p1}"`
  );
};

export const replaceDropdownTargetId = (htmlString) => {
  return htmlString.replace(
    /id="mantine-[a-z0-9]+-(dropdown)"/g,
    (match, p1) => `id="mantine-mockedId-${p1}"`
  );
};

export const replaceDataFocusId = (htmlString) => {
  return htmlString.replace(
    /data-focus-id="mantine-[a-z0-9]+"/g,
    'data-focus-id="mantine-mockedId"'
  );
};

export const replaceAriaLabelledBy = (htmlString) => {
  return htmlString.replace(
    /aria-labelledby="mantine-[a-z0-9]+-target"/g,
    'aria-labelledby="mantine-mockedId-target"'
  );
};

export const replaceAriaControls = (htmlString) => {
  return htmlString.replace(
    /aria-controls="mantine-[a-z0-9]+-dropdown"/g,
    'aria-controls="mantine-mockedId-dropdown"'
  );
};

export const replaceAllRandomGeneratedIds = (htmlString) => {
  let result = replaceTabId(htmlString);
  result = replaceMenuTargetId(result);
  result = replaceDropdownTargetId(result);
  result = replaceDataFocusId(result);
  result = replaceAriaLabelledBy(result);
  result = replaceAriaControls(result);

  return result;
};
