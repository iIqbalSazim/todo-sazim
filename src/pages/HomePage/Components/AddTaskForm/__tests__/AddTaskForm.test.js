import { render, screen } from "@testing-library/react";

import { replaceAllRandomGeneratedIds } from "../../../HomePageHelpers";
import AddTaskForm from "../AddTaskForm";

describe("AddTaskForm", () => {
  test("renders correctly", () => {
    const { container } = render(
      <AddTaskForm isOpen={true} closeModal={() => {}} />
    );

    const titleElement = screen.getByText("Create Todo");
    const taskInputElement = screen.getByPlaceholderText("Write your task");
    const setPriorityElement = screen.getByRole("searchbox");
    const dueDateElement = screen.getByRole("textbox", {
      name: /due date/i,
    });
    const createButtonElement = screen.getByRole("button", {
      name: "Create",
    });

    expect(titleElement).toBeInTheDocument();
    expect(taskInputElement).toBeInTheDocument();
    expect(setPriorityElement).toBeInTheDocument();
    expect(dueDateElement).toBeInTheDocument();
    expect(createButtonElement).toBeInTheDocument();

    expect(replaceAllRandomGeneratedIds(container.innerHTML)).toMatchSnapshot();
  });
});
