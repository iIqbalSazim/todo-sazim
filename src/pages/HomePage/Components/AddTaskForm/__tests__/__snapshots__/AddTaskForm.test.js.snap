// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTaskForm renders correctly 1`] = `""`;
